### DISCUSSION TITLE
How to fix: llama.cpp built without libcurl

### DISCUSSION THREAD

**Discussion by eonward** [ggml-org/llama.cpp#9835] [Q&A] [answered]
Hi! 

It seems like my llama.cpp can't use libcurl in my system. 

When I try to pull a model from HF, I get the following:
`llama_load_model_from_hf: llama.cpp built without libcurl, downloading from Hugging Face not supported.`

I'm on Ubuntu, and have the following modules installed:
`libcurl3t64-gnutls`
`libcurl4t64`

`libcurl4t64` in particular provides libcurl4
```
➜  llama.cpp git:(master) dpkg -L libcurl4t64
/usr/lib/x86_64-linux-gnu/libcurl.so.4.8.0
/usr/share/doc/libcurl4t64/changelog.Debian.gz
/usr/share/lintian/overrides/libcurl4t64
/usr/lib/x86_64-linux-gnu/libcurl.so.4
```
(abridged)

Is this the curl used in `common_download_file` in `common.cpp`, or is that looking for something else?

Any ideas? Thanks!

**Comment by danbev** [comment, marked as answer]
Can you double check that the llama.cpp version that you build used the `LLAMA_CURL` flag? 

If using cmake this would look something like this:
```console
$ cmake -S . -B build -DLLAMA_CURL=ON
```
Or if using make:
```console
$ make LLAMA_CURL=1
```

Also, if I remember correctly you might need the `libcurl4-openssl-dev` package as well.
```console
$ sudo apt install libcurl4-openssl-dev
```

**Reply by brauliobo** [reply]
worked here

**Reply by cryptonight-creator** [reply]
Worked for me too 😊😊

**Reply by jferments** [reply]
^ This works for Ubuntu 24.10. This should be in the installation instructions, or should be included in the error message when the library is missing.

**Comment by cyberluke** [comment]
Why this is not in BUILD.md ? And this seems to be required, not optional. Why this is not in Makefile? (latest github version 18.2.2025)

**Reply by danbev** [reply]
> Why this is not in BUILD.md ?

Currently not all build options have been documented in [build.md](https://github.com/ggml-org/llama.cpp/blob/master/docs/build.md) but ideally all will eventually. 

> And this seems to be required, not optional. 

It is required only if support for fetching models is needed, like downloading from huggingface for example. If you are running into a situation where this is not true then please open an issue so it can be tracked and looked into.

> Why this is not in Makefile? (

The Make build is deprecated in favor of CMake, but there a variable to handle [LLAMA_CURL](https://github.com/ggml-org/llama.cpp/blob/73e2ed3ce3492d3ed70193dd09ae8aa44779651d/Makefile#L1023) in the Makefile.

**Reply by cyberluke** [reply]
why you presume fetching models is not default feature?

**Reply by cyberluke** [reply]
The question was not only about Makefile, but also CMakefile. The issue is generic, not specific to Make or CMake.

**Reply by cyberluke** [reply]
I point your knowledge of build etiquette is completely wrong. LLAMA_CURL is stupid solution....but then again what one can expect from LLAMA, right? In that case best effort, great work!

**Reply by danbev** [reply]
> why you presume fetching models is not default feature?

I think this has to do with CuRL and that it might not be available on all systems that are supported.

**Reply by cyberluke** [reply]
1) nowadays CuRL is even on Windows
2) if that -> implement fail-over-solution for CuRL not avail

**Comment by tilkinsc** [comment]
Since this discussion didn't have any solutions one might be looking for on windows, I installed vcpkg
```batch
git clone https://github.com/microsoft/vcpkg.git
cd vcpkg
bootstrap-vcpkg.bat
```

Then I installed libcurl
```batch
vcpkg install curl:x64-windows
```

Then I integrated vcpkg so that libs can be automatically found
```batch
vcpkg integrate install
```

From there it emitted a cmake file I could pass to cmake via a -D
```
-DCMAKE_TOOLCHAIN_FILE=C:/code/vcpkg/scripts/buildsystems/vcpkg.cmake
```

And then I built llama.cpp
```batch
git clone https://github.com/ggerganov/llama.cpp.git
cd llama.cpp
cmake -B build -DLLAMA_CURL=ON -DGGML_CUDA=ON -DCMAKE_TOOLCHAIN_FILE=C:/code/vcpkg/scripts/buildsystems/vcpkg.cmake
cmake --build build --config Release
```

You might get away with using ninja, but I don't think it really matters since most of the time compiling is spent in .cu (cuda) files.

**Reply by Column01** [reply]
THANK YOU!

**Comment by Goutamxd** [comment]
**Hey @eonward , Your llama.cpp build lacks libcurl support, preventing Hugging Face downloads. 🚀 To fix this, install the required package:**

`` sudo apt update && sudo apt install `libcurl4-openssl-dev ``

**Then, rebuild llama.cpp with libcurl enabled:**

 `` LLAMA_CURL=ON`` make -j$(nproc) ``

**Check if libcurl is linked using ldd ./main | grep curl. ✅ If issues persist, ensure no conflicting versions (libcurl3t64-gnutls vs libcurl4t64) exist. Alternatively, download models manually via wget or git-lfs. Hope this helps! 🔥**
